core/autocrypt/claimchain.py

```diff
@@ -0,0 +1,223 @@
+        heads = self._getheads()
+        return heads.get(ident)
+
+    def _getheads(self):
+        if os.path.exists(self._path):
+            with codecs.open(self._path, "r", encoding="utf-8") as f:
+                return json.load(f)
+        return {}
+
+    def upsert(self, ident, cid):
+        heads = self._getheads()
+        heads[ident] = cid
+        with codecs.open(self._path, "w", encoding="utf-8") as f:
+            json.dump(heads, f, indent=2, sort_keys=True)
+
+
+class ClaimChain:
+    """ A claimchain maintains an append-only log where each entry
+    in the chain has its own content-based address so that claimchains
+    can cross-reference entries within each other.  Each entry in a chain
+    carries a timestamp and a parent CID (block hash) and type-specific
+    extra data.  Here are the current types and the contained extra data:
+
+    - genesis: cryptographic public key material.  parent must be None.
+
+    - oob_verification: identity, verified-head of this identity)
+
+    """
+    def __init__(self, blockservice, headtracker, ident):
+        self._bs = blockservice
+        self._ht = headtracker
+        self.ident = ident
+
+    def dump(self):
+        l = list(self.get_head_block())
+        for x in reversed(l):
+            pprint("{} {}: {}".format(x.timestamp, x.type, shortrepr(x.args)))
+
+    def append_block(self, type, args):
+        head = self.get_head_block(raising=False)
+        if head:
+            head = head.cid
+        block = self._bs.store_block(type, args, parent=head)
+        self._ht.upsert(self.ident, block.cid)
+        return block.cid
+
+    def get_head_block(self, ident=None, raising=True):
+        if ident is None:
+            ident = self.ident
+        head_cid = self._ht.get_head_cid(ident)
+        if not head_cid:
+            if raising:
+                raise ValueError("no head found for {}".format(ident))
+        else:
+            return self._bs.get_block(head_cid, raising=raising)
+
+    def iter_blocks(self, type=None):
+        for x in self.get_head_block():
+            if type is None or x.type == type:
+                yield x
+
+    def num_blocks(self):
+        return len(list(self.iter_blocks()))
+
+    def add_genesis(self, keydata):
+        assert not self.get_head_block(raising=False), "already have a genesis block"
+        assert isinstance(keydata, bytes)
+        ascii_keydata = mime.encode_binary_keydata(keydata)
+        self.append_block("genesis", [ascii_keydata])
+
+    def get_genesis_block(self):
+        head = self.get_head_block()
+        block = head.get_first_block()
+        assert block.type == "genesis"
+        return block
+
+    def add_oob_verify(self, email, cid):
+        assert self.get_head_block()
+        self.append_block("oob_verify", [email, cid])
+
+    def is_oob_verified_block(self, cid):
+        for block in self.iter_blocks(type="oob_verify"):
+            email, cid = block.args
```

Comment:
I think this overwrites /shields the cid passed in as an argument and thus the function will always return true.
At least i don't see how the output of the function would depend on the cid argument